python/cudf/cudf/core/frame.py

```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
+        columns = [
+            ([] if ignore_index else list(f._index._data.columns)) +
+            [f[name]._column if name in f else None for name in names]
+            for i, f in enumerate(objs)
+        ]
+
+        # Get a list of the combined index and table column indices
+        indices = list(range(functools.reduce(max, map(len, columns))))
+        # The position of the first table colum in each
+        # combined index + table columns list
+        first_data_column_position = len(indices) - len(names)
+
+        # Get the non-null columns and their dtypes
+        non_null_cols, dtypes = get_non_null_cols_and_dtypes(indices, columns)
+
+        # Infer common dtypes between numeric columns
+        # and combine CategoricalColumn categories
+        categories = find_common_dtypes_and_categories(non_null_cols, dtypes)
 
         # Cast all columns to a common dtype, assign combined categories,
         # and back-fill missing columns with all-null columns
-        for name in unique_columns_ordered_ls:
-            dtype = dtypes[name]
-            for df in objs:
-                # If column not in this df, fill with an all-null column
-                if name not in df:
-                    df[name] = column_empty(len(df), dtype, masked=True)
-                else:
-                    # If column is categorical, rebase the codes with the
-                    # combined categories, and cast the new codes to the
-                    # min-scalar-sized dtype
-                    if name in categories:
-                        df[name] = (
-                            df[name]
-                            .cat._set_categories(
-                                categories[name], is_unique=True
-                            )
-                            .codes
-                        )
-                    df[name] = df[name].astype(dtype)
-
-        # Arrange the columns from each DF in the same order
-        objs = [df[list(unique_columns_ordered_ls)] for df in objs]
+        cast_cols_to_common_dtypes(indices, columns, dtypes, categories)
+
+        # Construct input tables with the index and data columns in the same
+        # order. This strips the given index/column names and replaces the
+        # names with their integer positions in the `cols` list
+        tables = []
+        for i, cols in enumerate(columns):
+            table_cols = cols[first_data_column_position:]
+            table_names = indices[first_data_column_position:]
+            table = cls(data=dict(zip(table_names, table_cols)))
+            if 1 == first_data_column_position:
+                table._index = as_index(cols[0])
+            elif first_data_column_position > 1:
+                index_cols = cols[:first_data_column_position]
+                index_names = indices[:first_data_column_position]
+                table._index = cls(data=dict(zip(index_names, index_cols)))
```

Comment:
`cls` would be `DataFrame` here for example where I think you want to construct a `MultiIndex` instead of a `DataFrame`

Reply:
I did try to construct a MultiIndex first, but after checking `table.pyx` logic, that seemed unnecessary. These are temporary Tables constructed for the libcudf `concatenate` call below, and `table.pyx` indicates all it needs is a Table-like value as its `_index` in order to create the combined list of column_views.

Reply:
Do we have a test concatenating DataFrames with MultiIndexes? If not can we add one?

Reply:
As pointed out by @trxcllnt offline:
https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/tests/test_concat.py#L153
https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/tests/test_dataframe.py#L955-L998